### ISSUE TITLE
Moving Around Worn Items

### ISSUE THREAD

**Issue by tajmone** [AnssiR66/AlanStdLib#60]
The problems surrounding transferall of worn items are starting to grow in number, to the point that it's worth having an issue in its own right dedicated to this topic — so that we might track all the affected verbs until the problem is solved.

The test transcript for this category of problems is now:

- [`tests/clothing/actions_transferring-worn-items.a3log`][actions_transferring-worn-items.a3log]

# Problem Description

Items being worn by an actor could be caused to be removed from its owner by a number of verbs, many of which are currently failing to handle changing the status of the item regarding its being `donned`, in the `wearing` of the actor, and (in case of the Hero) being inside `worn`.

Library verbs that could affect worn items are (column "err" indicates whether the verb/action fails to handle worn items, "???" indicates not tested):

|   verb  | err |                      description                       |
|---------|-----|--------------------------------------------------------|
| `take`  | ??? | Compliant NPCs will give the worn item to the Hero.    |
| `ask`   | YES | NPCs always give the worn item to the Hero.            |
| `put`   | YES | Hero can put worn item in a container.                 |
| `drop`  | ??? |                                                        |
| `throw` | ??? | Authors might allow throwing things, must keep an eye! |

And I'm sure there are more verbs missing still.

# Layered Clothing Issues

Another problem came up during tests: verbs that move a worn item to another place (container, floor, an NPC) are not taking into account if the worn items is blocked by an upper layer clothing — which breaks the logic of layered clothing in the library.

For example (from the same transcript):

```
> DBG boxers
'boxers' VALUES: | botcover: 2 | 
DONNED: Yes 
IN WORN: Yes 
IN WEARING OF: | hero

> put boxers in basket
You put your boxers into the basket.

> ; ** ERROR!!! ** ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ;                The boxers are worn by Hero, underneath his trousers, so he
> ;                shouldn't be able to just put them anywhere before first
> ;                REMOVING trousers AND THEN boxers!
> ;                This breaks right through the whole logic of layered clothing!
> ;                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This type of problem could affect all the verbs which allow moving around object (mentioned above).

# General Considerations

I think that we should find an elegant solution to these problems, taking into account that authors will be creating their own verbs for handling `thing`s and `object`s, which will also affect worn clothing. So, it would be easy for authors (especially newbies) to overlook the implications these verbs might have on clothing items being worn (by Hero or NPCs).

Patching the library verbs would fix the current problems but not ensure a problem-free usage of the library when custom verbs are introduced.

Every turn events would be better to handle and fix status of worn clothes that have been detached from their wearer.

The issue of preserving consistency of layered clothing is more tricky though. How are we to enforce on all such verbs checking that a worn item is not blocked by another item above it?

The library won't let you remove your vest if you're wearing a pullover, but it allows you to drop it, give it to someone, put it in a box, etc., without having to remove the pullover first. This kind of breaks the whole purpose and logic of layerd clothing-

As a fact, it might easier to just drop a vest instead of having to remove a t-shirt, a shirt, a pullover and a coat first, and then having to wear them all back on again! But this type of "shortcuts" might break a whole adventure, e.g. if some items were not supposed to be removable (e.g. when handcuffed) by some blocks on the `remove` verb (which would be bypassed by this loophole).

[actions_transferring-worn-items.a3log]: https://github.com/AnssiR66/AlanStdLib/blob/clothing/tests/clothing/actions_transferring-worn-items.a3log

**Comment by AnssiR66** [comment]
Great that you were able notice these bugs in the tests Tristano. I think it all mainly boils down to the fact that we have to limit the implicit transfer of clothes when they are worn by anyone.
Because the library handles layered clothing correctly when putting it on or removing it, only implicit actions must be addressed.
So, would EXTRACT work here? :

EVERY CLOTHING ISA OBJECT...
CONTAINER...
EXTRACT
    CHECK THIS IS NOT donned
       ELSE SAY THE THIS. "would have to be taken off first."
...
END EVERY.

I'll try this too, shortly. But yesterday, for some reason, I could not make EXTRACT work ...
Alternatively, this limitation can be easily added to every verb having an implicit action in lib_verbs.i.
My ultimate point being: we must prevent clothing items from being handled implicitly at all when they are worn by someone.

-Anssi





________________________________

**Comment by tajmone** [comment]
# Introduce a `blocked` Attribute

Maybe we could add a new `blocked` attribute to the `clothing` class, which the library can then set to true/false during wear/remove operations.

This attribute would allow a simple `Check` to be added to any verb that might move around clothing, by adding the same verb on the `clothing` class, just to enforce those checks and make the action fail if the item is `blocked`.

This solution would be easier to handle (even by authors addin new verbs) than having to do all the calculation of every item in the wearing of the ACTOR — this attribute being handled automatically in the `wear` and `remove` verbs only, and being basically a "read-only" attribute in all other contexts.

So, roughly, we could redefine `put` in the clothing class as:


```alan
  VERB put_in
    WHEN obj
      CHECK obj NOT blocked
        ELSE "You'd need to remove the clothing on to of $+1 first!" 
```

And probably this could apply to multiple verbs with similar SYNTAXes, effectively requiring only a couple of these to cover every verb that could potentially affect worn and blocked items.

What do you think?

**Comment by tajmone** [comment]
Good idea. But shouldn't we targetting ACTORs here? After all a clothing item can only be worn inside an actor, or the `worn` in case of the Hero:

And here we could also apply the `blocked` attribute I proposed:

``` alan
EVERY CLOTHING ISA OBJECT...
  CONTAINER...
  EXTRACT
    CHECK THIS IS NOT blocked --> Only prevent blocked items?
      ELSE "You'd need to remove the clothing on to of $+1 first!" 
...
END EVERY.
```

Surely, a blocked item (which is always a _worn_ item also!) should never leave an NPC.

It's not clear to me though how this EXCTACT on CLOTHING works. What does the THIS refer to in the above CHECK, the actual clothing item or its container? It would be useful if we could apply the CHECK to ACTORS instead, but be able to verify therein attributes of the item being extracted. The documentation isn't really cleary on this.

I mean ... ACTORS are CONTAINERS, but since they can't be _inside_ containers what does THIS refer to in the EXTRACT of an ACTOR? I'm a bit confused here...

Anyhow, because the clothing class is so pervasive in the library, we _really_ have to consider carefully all the possible approaches, and which one would be the best one in terms of how it impacts authors custom verbs (IMO).

**Comment by tajmone** [comment]
@thoni56,

Could you please help me understand better how `EXTRACT` work?

If I've understood correctly, it's only applicable to `CONTAINER` types, right?

In order to readjust the StdLib clothing to work around the current bug occuring when an `EVENT` triggers an `EXTRACT` we have to be careful in the new approach.

What is not clear to me is whether `EXTRACT` can reference both the container and the object being extracted. I assume that `THIS` inside an `EXTRACT` block will be refering to the container from which something is being extracted. But how could I, for example, conditionally evaluate the class or attributes of the object being extracted? Is it possible?

The confusing part here is that ACTORs are containers, but they can't be inside containers, so the meaning of `EXTRACT` is actually quite different inside an ACTOR and an OBJECT.

Clearly, using `EXTRACT` to ensure that a worn item being moved around is set as no longer `donned`, and removed from the `wearing` of any ACTOR is the easiest way to provide a library approach to this problem that won't burden end-authors with having to handle this in their new custom verbs.

My idea is that, if it's possible to add an `EXTRACT` to the ACTOR class, able to check if the item being removed is of `clothing` type, then we won't need to worry about verbs having to deal with this. The same goes for the `worn` entity, which holds the Hero's worn clothing and would need an `EXTRACT` too. This approach would make usage of EVENTs superfluos, and we'd be avoiding the bug altogether.

**Comment by tajmone** [comment]
I find EXTRACT confusing. I'm doing some tests in a pure Alan file.

I've created an ACTOR and added an EXTRACT to it:

```alan
THE man IsA actor AT room.
  CONTAINER
  EXTRACT
    "-------------$n"
    "Extraction$n"
    "THIS:" SAY THIS. ".$n"
    "$$o: $o"
    "$n-------------$n"
END THE.
```

When I take a ball object held by the man, the EXTRACT above show that `THIS` refers to `man` and `$o` prints `ball`. 

But beside using the `$o` in output, I can't refer to the object being extracted by using `o` (Alan's default for implicit verb parameter name), nor `obj` (the actual parameter as defined in my custom `take` verb).

So, it's possible to reference the extracted object in print but not in conditional code checks. (at least, I haven't found a way to do it).

The documentation doesn't say much on this, beside stating that CHECK and DOES behave like in VERBs.

**Comment by AnssiR66** [comment]
@Tristano, 
as I understand it, EXTRACT defines what happens when something is taken out of that container. It is a restricting statement limiting the removing of items from the container. 

THE man ISA ACTOR
CONTAINER
   ...
   EXTRACT
       CHECK man IS asleep
         ELSE "You can't take anything from him while he is awake."

The CHECK part lists the exceptional situations when taking /removing actually is possible, ELSE defines the failing message. 

Now, because Worn is a container, I think we could define the Extract for clothing items to be there. That would take care of quite many things if it is successful. But how to define it?

THE worn ISA ENTITY
Container
...
EXTRACT
        CHECK ???
             ELSE "You should remove that first."
                          -- 'that' referring to the piece of clothing
...
END THE.

Then, we should find a way to just limit 'blocked' items from being removed, and 'NOT blocked' items to be transferred freely.
The 'remove' verb should also declare
"DOES MAKE THIS NOT blocked.
LOCATE THIS IN hero."
etc.

Just some preliminary thoughts...  I'll write back

**Comment by AnssiR66** [comment]
I am actually not sure how extract is supposed to work. I tried both of these, and both compiled fine:

THE worn ISA ENTITY 							
	CONTAINER TAKING CLOTHING.			
		HEADER SAY hero_worn_header OF my_game.
		ELSE SAY hero_worn_else OF my_game.
		EXTRACT
			CHECK "You should remove $o+1 first."     --unconditional check, never allows removal
END THE.

and
THE worn ISA ENTITY 							
	CONTAINER TAKING CLOTHING.			
		HEADER SAY hero_worn_header OF my_game.
		ELSE SAY hero_worn_else OF my_game.
		EXTRACT
			"You should remove $o+1 first."               -- without CHECK
END THE.

but neither limited the giving action; I was able to give to an NPC a worn clothing item...

**Comment by tajmone** [comment]
I find myself confused just as you. 

I _think_ the problem here with "`$o`" is that — although it does work in output — it's just dragging along the reference from the VERB which is causing this EXTRACT to occur, but since trying using "`$1`" causes a compiler error stating that parameters are not available in this EXTRACT context, my guess is that this is due to the fact that "`$o`" is a now depracted reference (it works, but probably it's not safe, and maybe shouldn't even be working at all).

As for your proposal of using EXTRACT in the `clothing` class, I can't see many benefits in this for it would be only exerting checks/limitations on items inside a trousers pockets (for example), which is not directly relevant to the problem at hand (these would/should be carried items anyhow).

Worn items will always be in ACTORs (NPCs) or `worn` (if Hero). But unless code inside an EXTRACT can access properties of the extracted item, it's usefulness is going to be limited. The documentation on EXTRACT doesn't say much on this topic.

My guess is that, like you said, the original intention of EXTRACT was (and still is) to be able to prevent a container giving up its contents based on some attributes of the container (a non coopeartive actor, a boling pot which the player couldn't be handling, etc.). So, probably all checks are aimed in that direction too — having at disposal only `THIS`, pointing to the COINTAINER hosting the EXTRACT.

All that I can think of is that ACTORs have the `wearing` set property, which can be accessed from within EXTRACT via `IN THIS:wearing`, but still I don't see how we could use this if we have no way of knowing what is being extracted — also, the Hero's `wearing` is disjointed from the `worn` entity, so this would only be accessibile in NPCs' EXTRACT too.

**Comment by AnssiR66** [comment]
@tajmone: You wrote
>>As for your proposal of using EXTRACT in the clothing class, I can't see many benefits in this for it would be only exerting checks/limitations on items inside a trousers pockets (for example), which is not directly relevant to the problem at hand (these would/should be carried items anyhow).

Yes, that's why EXTRACT should be in the Worn container, not in the Clothing class. But for the reasons you stated, I find it difficult  to formulate right CHECK there (like I demonstrated above), and furthermore, I cannot seem to make the CHECK to work anyway....

As Worn is a container that just takes clothing (Container Taking Clothing), and then, when there is a wallet in a jacket pocket, for example, this wallet is indirectly in Worn. But, as you found out earlier, seems like the wallet is treated like a clothing item, or as if it was directly in Worn,  because of the message to remove it first... I didn't encounter this problem before, so I keep testing...

**Comment by AnssiR66** [comment]
OK, now at least something is working:
1) I deleted the check_worn_clothing event, and all references to it
2) I declared the worn entity as:

THE worn ISA ENTITY 	
	IS blocked.						
	CONTAINER TAKING CLOTHING.			
		HEADER SAY hero_worn_header OF my_game.
		ELSE SAY hero_worn_else OF my_game.
		EXTRACT 
			CHECK THIS IS NOT blocked
			ELSE "You should take off the $o first."
END THE.

3) I changed the last lines of the DOES part of the 'remove' verb:

                       MAKE worn NOT blocked.
			LOCATE THIS IN hero.
			"You take off" SAY THE THIS. "." 
			EXCLUDE THIS FROM wearing OF hero.
			MAKE THIS NOT donned.
			MAKE worn blocked.
	END IF.

This amounts to:
- I cannot implicitly give worn items, which is what I think is the best solution; I have to remove them first
- I can reference the clothing item with $o in the extract message 

EXTRACT seems  to work now the way it is supposed to.


Currently I am working on why clothing taken from an NPC goes to worn, not just to the hero in general.

**Comment by tajmone** [comment]
> I used 'blocked' to the whole worn entity, not individual clothing, because I think we can be consistent and disallow all clothing from being moved implicitly when it's worn. 

Well though Anssi, I didn't think of reversing the logic of `blocked` by attaching it to the `worn` instead. I also think that having the main `take` (in `lib_verbs.i`) handle clothing is easier to follow and maintain that overriding it on the `clothing` class.

> Would this be too limiting?

As a gut response, I'd say that this should be norm and anything else the exception. In most cases the player should be aware that removing an item is required before doing something else with it. If for some actions it would make sense to carry out an implicit clothing removal, then we could just add some code to their verbs. 

But in general, I'd say that a strict behaviour for this is better, especially in view of author created verbs, which should default to not being able to affect worn items, leaving it to authors to otherwise allow this if they feel it's needed. Strictness would prevent the library falling apart in big adventures with lot's of new verbs, I think. But of course, only real life testing can truely answer this (and the test suite might help here).

> As you said, Tristano, that tempworn could be got rid of, this might be another good reason for doing it. Do you think it would not affect matters if we just scrapped it and programmed the "You are already wearing" part in some other way?

In the Italian library I've already scraped it off and everything is working fine. For the "You are already wearing" part I'm still using a simple `LIST worn`, which has the same effect but without the "(being worn)". So, probably there is not need to add new code for handling lists of worn clothing in actors. Did you have in mind, some new approach to this?

# Start Using Dev Snapshots?

A major question here is if we should keep testing the StdLib with the latest official Alan build or start using dev snapshots instead.

In the Italian library I'm using dev snapshots (I have to, because features for Italian accents support are only available in dev snaps), while in the StdLib `tests/` folder I've pasted the binaries of Alan 3.0 beta6 to ensure it runs against official releases.

These are the pros of using dev snapshots instead of releases:

- We'd be running tests with latest bug fixes, making the StdLib ready for the upcoming beta7.

And these the cons:

- We might be developing assuming some fixes and features not yet in official releases, without realizing, so the StdLib might not work as expected when end users use official releases — but this wouldn't be a problem if we start using strict versioning of the StdLib and work on dev branches and only merge into `master` when beta7 is out (we could then even stop using snapshots).
- We must always be in synch with each other as to which version we are using to run tests, to avoid different results, and always test, discuss and agree on moving to a newer dev snapshot (these being quite frequently updated). 

For the last point, we could even add the SKD binaries (Windows) directly to the project, in the `tests/` folder — usually is not a good practice to drop binaries in Git projects, but sometimes you just have to do it (eg. DLLs and static libraries which are complicate to compile, etc.). Size wouldn't really be a great issue, these binaries are small (although ultimately, there'd be many of them in time).

I personally think that a library should always be made publicly available for latest official version of a language, but Alan being still in beta this might not be the case here. Also, I'm speaking here about the _tests suite_ only — which binaries end users employ to create their adventure is their own choice.

But we have to decide a line of work here, for the dev snapshots have fixed a good number of bugs since beta6, which also include PRONOUNs, and other features which might be of interest to the StdLib; besides, Thoni has been actively working on the EXTRACT bug in the last days, and in the latest dev snapshots the bug has already changed in nature. 

Anyhow, ultimately the most important thing is that we (as _me_ and _you_) are both using the same Alan version binaries for the tests.

**Comment by AnssiR66** [comment]
I think that for the reasons you stated, it would be good to use the latest dev snapshots...shall we do that?

And ok, we can go with the worn container having the blocked attribute, and if the player wants to allow implicit handling of certain clothing items, we can figure out something (DOES ONLY, or something else).

About the DOES part of the 'take' verb: actually, even if you mentioned it is fine if the section that deals with the clothing is in the general 'take' verb (lib_verbs.i), it might be more congruous with the rest of the library if it after all is placed in the classes.i file under clothing.... because it makes the DOES part of the general 'take' verb much clearer, and anyway the code dealing with clothing there is kind of background workings that could be placed in the clothing class, "out of the way".... 

(Lastly, a noobie question (I'm still not comfortable with repositories..:) .. When I try to download the repository files (the five lib_ ... files ) to my own local folder, the result is gibberish when I try to open these files in WordPad... any quick hint as to what to do? Of course I can can copy/paste the text, but would there be a more straightforward way, of just downloading the files straight, without it getting corrupt..? Probably I'm missing something obvious...)

**Comment by AnssiR66** [comment]
Never mind my last question, ..I found already the "clone/download" button; I had tried to download the files individually :D

**Comment by tajmone** [comment]
> I think that for the reasons you stated, it would be good to use the latest dev snapshots...shall we do that?

I too think is a good idea. I have the latest snapshots on my system PATH already, and had placed the beta6 binaries specifically in the StdLib `tests/` folder to ensure that tests with the English library would use the latest official release.

The I'll just add in the README document the snapshot version being used, so that anyone can always check there for the required (advised) version — after all, we won't be able to switch to _every_ new dev snapshot, for these are quite frequent, and also we need to run all tests to make sure that new changes don't break the library code.

> About the DOES part of the 'take' verb ... it might be more congruous with the rest of the library if it after all is placed in the classes.i file under clothing.... because it makes the DOES part of the general 'take' verb much clearer.

In hindsight, I had come to your same conclusion. Although it wouldn't be too big of a problem, separation is always better. 

# Alan Italian: Making Clothing Optional

In the Italian library one of my goals is to ultimately make the clothing an optional feature — as a fact, right this morning I've mode all clothing code to a submodule (something like `lib_classes_clothing.i`). Probably implementing a fully "plugin" optional module for clothing won't be easy, but I'll try. I think that not imposing an advanced feature like clothing would be good, for many adventures won't be interested in using it, other might want to replace it with a custom system. Freedom of choice is always better than imposition (but modularity can get complex in Alan).

# Working Locally With StdLib and Git

> When I try to download the repository files (the five lib_ ... files ) to my own local folder, the result is gibberish when I try to open these files in WordPad... any quick hint as to what to do?

The problem is going to be that all Alan related files are in ISO-8859-1 (Alan sources, but also tests scripts/logs), and most code editors handle _very badly_ ISO-8859-1 (which is not really used by many people today). Chances are that whenever you paste something into the source files the editor will switch the encdoing to UTF-8, __breaking the shole source__.

When I started to work on the Italian library I went through so much pains because of  ISO-8859-1 that I decided to create my Sublime Alan package to solve all these problems, and defined some strict syntaxes for Alan sources, as well as commands scripts and transcripts (this being the reason I also needed to add custom extensions to these).

Basically, unless you can configure your editor to never break the encoding (ie. must reject Unicode/UTF-8 pasted text), you'll be facing some serious headhackes, especially since Git is not really aware of ISO-8859-1, so it won't do much in this respect.

My advice is that you try to use Sublime Text and my Sublime Alan package:

- https://www.sublimetext.com/
- https://github.com/tajmone/sublime-alan 

You can freely install the latest version of Sublime Text (but not the developer builds, which are for licensees only) and keep using it without features restrictions until you feel comfortable deciding whether you like or not, before buying a license. Since editors take long to learn and get used to, the creators of Sublime Text chose not to impose time limts, and allow potential customers to take all the time in the world to decide if it's the right editor for them (eg., someone might have only a day a week to play around with it, others an hour a day, ecc.). I have used it for many monthes before buying the license, i.e. until I felt comfortable switching all my work to ST from my previous editor.

Sublime Text now offers good native integration with Git too, so it's worth trying it. I also use Sublime Merge as a Git GUI, which is the companion of Sublime Text and they work in tandem marvellously:

- https://www.sublimemerge.com/

With Sublime Alan installed, you'll also get some syntax highlighting in Git diffs, and syntax awareness for Alan encodings (carried over from ST).

> Of course I can can copy/paste the text, but would there be a more straightforward way, of just downloading the files straight, without it getting corrupt..? Probably I'm missing something obvious...)

**_Don't_**! the encoding issue is too delicate, and broken encodings just go by unnoticed!

> I found already the "clone/download" button; I had tried to download the files individually :D

**_Don't_** use the "clone/download" button either! it only provides Linux based clones of the repository, which affects EOL normalization (skips native EOL on Windows) and, depending on your Git settings, could really mess up things. Same thing with raw files downloading: EOL normalization is handled as in Linux.

Also, I strongly advice you enforce some safety settings on Git regarding EOL normalization:

    core.eol = native
    core.autocrlf = true
    core.safecrlf = true

(See the [_Mind the End of Your Line_](https://adaptivepatchwork.com/2012/03/01/mind-the-end-of-your-line/)) article on this).

I wish it was simpler, but having gone through the whole ordeal when I first embraked on the Italian translation, I've learned the lesson and resorted to creating my Sublime Alan — encoding is no longer an issue, and it allows me to work _really_ fast, and handle autocompletions and snippets for both StdLib EN/IT, and handle game scripts/logs too. Unfortunately, time being an issue, speed of editing is a top requirement in my translation of Alan Italian, and every saved second counts on the long term.

> And yes, let's get rid of tempworn... Tristano, do you mind doing it, the same way as what you have done with the Italian library, so it will be in sync...

I will!

**Comment by AnssiR66** [comment]
Good! 
Good! And thanks also for explaining me about the downloading of files etc.

I have now fixed everything I can find at fault about clothing; maybe I have still overlooked something but we can take a further look if needed. I would shortly commit these changes to the repository, but I list them here:

1) Adding the Extract section to the Worn container, and giving the Worn container the 'blocked' attribute:

THE worn ISA ENTITY 	
	IS blocked.						
	CONTAINER TAKING CLOTHING.			
		HEADER SAY hero_worn_header OF my_game.
		ELSE SAY hero_worn_else OF my_game.
		EXTRACT 
			CHECK THIS IS NOT blocked
			ELSE "You should take off the $o first."
END THE.

2) Deleting the event 'worn_clothing_check' altogether, and any references to it

3) Adding the verbs take and take_from to the clothing class, to ensure that 'donned' and 'wearing' attributes are updated correctly:


VERB take
	DOES ONLY
		IF THIS DIRECTLY IN worn
			THEN 
			
				MAKE worn NOT blocked.
	
				EXCLUDE THIS FROM wearing OF hero.
				MAKE THIS NOT donned.
				LOCATE THIS IN hero.
				"You take off" SAY THE obj. "and carry it in your hands."	

				MAKE worn blocked.			
			
			ELSE 
				FOR EACH ac ISA ACTOR DO
					IF THIS IN wearing OF ac
						THEN EXCLUDE THIS FROM wearing OF ac.
							MAKE THIS NOT donned.
							"(from" SAY THE ac. "$$)$n"
					END IF.			
				END FOR.
				LOCATE THIS IN hero.
			 	"Taken."
		END IF.

END VERB.
		

VERB take_from
	WHEN obj
	DOES ONLY

		FOR EACH ac ISA ACTOR DO
			IF THIS IN wearing OF ac
				THEN EXCLUDE THIS FROM wearing OF ac.
					MAKE THIS NOT donned.
			END IF.			
		END FOR.
		
		LOCATE THIS IN hero.
		"Taken."
		

END VERB.

4) So that we can take objects from worn clothing items, I had to add anyway this coding to 'take' in lib_verbs.i:

(take) :

DOES
			IF obj ISA ACTOR
				THEN SAY THE obj. "would probably object to that."
			
			ELSIF obj ISA OBJECT
				THEN 
					IF obj IN worn       -- to allow items in clothing items to be taken
						THEN MAKE worn NOT blocked.
					END IF.

					LOCATE obj IN hero.
					"Taken."

					IF worn IS NOT blocked   
						THEN MAKE worn blocked.
					END IF.

			END IF.

This had to be done, because for example a wallet in a jacket is indirectly in Worn, and the Extract prevents it from being taken if we don't allow that temporary unblocking of Worn.

Now, everything should work, including the unnecessary "being worn" not showing. I would go ahead and make these changes to the repository, if you don't see anything preventing that at this point?

Also, you took up the descriptions of 'hero' and  NPCs as far as carrying and wearing items is concerned.
I did originally (a year of two ago) try to make them identical, but it turned out very difficult to have an NPC's clothing placed in something like the Worn for the hero; it was just not practical. Because for the hero it is simple enough: LIST hero. LIST worn. But what should we list for an NPC? The author should always coin up separate containers for each NPC's clothing, and then code the clothing to be located there manually. That's just not practical... we can sooner make the hero have the ("being worn") method of listing, if we want to have these two methods the same (and if I remember correctly, that's what Inform does, as well, for example). I understand that the two existing methods (for the hero and for an NPC) in Alan are not identical at this stage, and it might be a bit confusing, or unsatisfactory. We can still think about this further.

**Comment by tajmone** [comment]
> Also, you took up the descriptions of 'hero' and NPCs as far as carrying and wearing items is concerned. I did originally (a year of two ago) try to make them identical, but it turned out very difficult to have an NPC's clothing placed in something like the Worn for the hero; it was just not practical.

I see. I haven't actually had a go at this, but what I had in mind was to tweak the `HEADER` of ACTORs (since they are containers) and try to use two different loops to list items therein, one for non-clothing (carrying) and another one after for clothing (worn). I.e. I was assuming this was doable by not using `LIST` bu custom loops.

Anyhow, it's not a big deal at all, and the Hero is always a special case anyhow in most things in an adventure, so it's not like anyone would see as strange. My proposal was more out of concern that depending on the order items are acquired, their listing might be a mixture of alternatig carried and worn, whereas a cleaner separation would have been nicer (but I guess reordering them could be done via some Event triggered at certain occasions, looping through items via a temporary container so they are sorted neatly).
